### ISSUE TITLE
Trouble Setting Conflict for Darwin aarch64/M1

### ISSUE THREAD

**Issue by ax3l** [LLNL/spack#26520]
### Steps to reproduce

In #26318, I want to set a conflict that GCC does not build on Apple M1 (Darwin aarch64) architectures prior to the GCC 12+ releases.

This works:
* `conflicts('@:11', when='platform=darwin')`

and those should as well, but don't:
* `conflicts('@:11', when='target=aarch64: platform=darwin')`
* `conflicts('@:11', when='target=m1 platform=darwin')`
* `conflicts('@:11', when='target=m1: platform=darwin')`

### Error message

N/A

### Information on your system

Same system as reported in #26317

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have searched the issues of this repo and believe this is not a duplicate
- [X] I have run the failing commands in debug mode and reported the output

**Comment by alalazo** [comment]
I started looking into this. First finding is that on `linux` with `x86_64` arch it seems to work, i.e. if I do: 
```
conflicts('@:11', when='target=x86_64: platform=linux')
```
`gcc` is correctly concretized at version `master`. Using the first conflict posted above for darwin generates these facts
```console
condition(604).
condition_requirement(604,"node","gcc").
condition_requirement(604,"version_satisfies","gcc",":11").
conflict_trigger(604).
condition(605).
condition_requirement(605,"node","gcc").
condition_requirement(605,"node_platform","gcc","darwin").
condition_requirement(605,"node_target_satisfies","gcc","aarch64:").
conflict("gcc",604,605).
```
which also seems correct. The expansion of `node_target_satisfies` facts are:
```
node_target_satisfies("gcc","aarch64:","aarch64").
node_target_satisfies("gcc","aarch64:","thunderx2").
node_target_satisfies("gcc","aarch64:","a64fx").
node_target_satisfies("gcc","aarch64:","graviton").
node_target_satisfies("gcc","aarch64:","graviton2").
node_target_satisfies("gcc","aarch64:","m1").
```
where `m1` is listed. Even `version_satisfies` facts seem correct:

<details>

<summary> version_satisfies </summary>
```
version_satisfies("gcc",":11","4.5.4").
version_satisfies("gcc",":11","4.6.4").
version_satisfies("gcc",":11","4.7.4").
version_satisfies("gcc",":11","4.8.4").
version_satisfies("gcc",":11","4.8.5").
version_satisfies("gcc",":11","4.9.1").
version_satisfies("gcc",":11","4.9.2").
version_satisfies("gcc",":11","4.9.3").
version_satisfies("gcc",":11","4.9.4").
version_satisfies("gcc",":11","5.1.0").
version_satisfies("gcc",":11","5.2.0").
version_satisfies("gcc",":11","5.3.0").
version_satisfies("gcc",":11","5.4.0").
version_satisfies("gcc",":11","5.5.0").
version_satisfies("gcc",":11","6.1.0").
version_satisfies("gcc",":11","6.2.0").
version_satisfies("gcc",":11","6.3.0").
version_satisfies("gcc",":11","6.4.0").
version_satisfies("gcc",":11","6.5.0").
version_satisfies("gcc",":11","7.1.0").
version_satisfies("gcc",":11","7.2.0").
version_satisfies("gcc",":11","7.3.0").
version_satisfies("gcc",":11","7.4.0").
version_satisfies("gcc",":11","7.5.0").
version_satisfies("gcc",":11","8.1.0").
version_satisfies("gcc",":11","8.2.0").
version_satisfies("gcc",":11","8.3.0").
version_satisfies("gcc",":11","8.4.0").
version_satisfies("gcc",":11","8.5.0").
version_satisfies("gcc",":11","9.1.0").
version_satisfies("gcc",":11","9.2.0").
version_satisfies("gcc",":11","9.3.0").
version_satisfies("gcc",":11","9.4.0").
version_satisfies("gcc",":11","10.1.0").
version_satisfies("gcc",":11","10.2.0").
version_satisfies("gcc",":11","10.3.0").
version_satisfies("gcc",":11","11.1.0").
version_satisfies("gcc",":11","11.2.0").
```

</details>

So at the moment I have troubles reproducing this issue.

**Comment by ax3l** [comment]
So I just tried again the merged version from #26318 with the latest Spack `development`.

If I am performing a `spack install gcc` at an AppleM1, it will automatically select `gcc@master`, which is ok as we unconstrained it.

If I remove the `master` version, I would expect a conflict now.
I kind-of get one, but not with the usual conflict message. I get "Error: gcc does not satisfy unknown" wand then a list of unsatisfiables, which is probably ok?

When I opened the issue, the constrain was simply ignored.
